Add Rotation control system

knelli2 opened the pull request:
## Proposed changes

Adds a rotation control system that controls the 3D quaternion rotation map.

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

knelli2 committed:
Add Rotation control system

Controls the 3D rotation map that uses quaternions

knelli2 committed:
Test Expansion control system names

wthrowe commented on src/ControlSystem/ControlErrors/Rotation.hpp:
```diff
@@ -0,0 +1,96 @@
+                        const tuples::TaggedTuple<TupleTags...>& measurements) {
+    const auto& domain = get<domain::Tags::Domain<3>>(cache);
+
+    using horizon_label = ::ah::ObjectLabel;
```
CamelCase because this is used as a normal type.

wthrowe commented on src/ControlSystem/ControlErrors/Rotation.hpp:
```diff
@@ -0,0 +1,96 @@
+    const DataVector grid_position_of_B = array_to_datavector(
+        domain.excision_spheres().at("ObjectBExcisionSphere").center());
+    const DataVector current_position_of_A = get<center_A>(measurements);
+    const DataVector current_position_of_B = get<center_B>(measurements);
```
References for both of these.

wthrowe commented on src/ControlSystem/ControlErrors/Rotation.hpp:
```diff
@@ -0,0 +1,96 @@
+ * positions of the centers of the mapped objects, and
+ * \f$\vec{C}\f$ is the difference of the centers of the excision spheres, all
+ * in the grid frame. It is assumed that the positions of the excision spheres
+ * are exactly along the x-axis.
```
You don't assume this in the code.

wthrowe commented on src/ControlSystem/Systems/Rotation.hpp:
```diff
@@ -0,0 +1,112 @@
+
+  using measurement = ah::BothHorizons;
+  static_assert(tt::assert_conforms_to<measurement,
+                                       control_system::protocols::Measurement>);
```
Just `tt::conforms_to`.

wthrowe commented on src/ControlSystem/Systems/Rotation.hpp:
```diff
@@ -0,0 +1,112 @@
+
+  using control_error = ControlErrors::Rotation;
+  static_assert(tt::assert_conforms_to<
+                control_error, control_system::protocols::ControlError>);
```
Again `tt::conforms_to`.

wthrowe commented on tests/Unit/ControlSystem/ControlErrors/Test_Rotation.cpp:
```diff
@@ -0,0 +1,126 @@
+
+  const DataVector expected_control_error = grid_cross_pos / grid_dot_pos;
+
+  CHECK(control_error == expected_control_error);
```
This test is a near copy-paste of the source being tested.  Instead of using random numbers and computing the expected answer in code, use specific inputs with a known result.

wthrowe commented on tests/Unit/ControlSystem/ControlErrors/Test_Rotation.cpp:
```diff
@@ -0,0 +1,126 @@
+
+  // Global things
+  domain::FunctionsOfTime::register_derived_with_charm();
+  double initial_time = 0.0;
```
`const`

wthrowe commented on tests/Unit/ControlSystem/Systems/Test_Rotation.cpp:
```diff
@@ -0,0 +1,186 @@
+
+  // Global things
+  domain::FunctionsOfTime::register_derived_with_charm();
+  double initial_time = 0.0;
```
`const`

knelli2 committed:
Small tweats to expansion system and tests

- Use tt::conforms_to_v instead of tt::assert_conforms_to in Expansion
system
- Forgo a type alias in Expansion control error
- Use known values in test of Expansion control error
- Make some variables const in Expansion tests

knelli2 commented:
Rebased on develop and added fixups. @wthrowe Also applied some of the changes you requested here to the Expansion control system as well in a new commit.

wthrowe reviewed (dismissed):
Go ahead and squash.  Thanks for cleaning up the expansion code too.

nilsdeppe commented:
@markscheel would you mind taking a look at this? Your expertise would be really useful here!

markscheel commented on src/ControlSystem/Systems/Rotation.hpp:
```diff
@@ -0,0 +1,112 @@
+ * map. It does this by directly controlling the internal PiecewisePolynomial in
+ * a \link domain::FunctionsOfTime::QuaternionFunctionOfTime
+ * QuaternionFunctionOfTime \endlink, which represents the angle the system has
+ * rotated through about each axis.
```
The 'Indirectly' and 'directly' phrases here are a bit confusing to me, because they made me think that either I should find some code in `Rotaton` that messes with the internals of QuaternionFunctionOfTime, or the user needs to be careful with where QuaternionFunctionOfTime is specified and where QuaternionFunctionOfTime's internal FunctionOfTime is specified.  But (I think) when the rotation control system is used, the user should simply use it with a QuaternionFunctionOfTime, and the user never needs to know that the QuaternionFunctionOfTime internally holds another PiecewisePolynomial.

So maybe this comment should be changed to something like:
"Controls the quaternion in the 3D Rotation coordinate map.  From the user's perspective, `Rotation` simply controls and updates a QuaternionFunctionOfTime.  (Internally, QuaternionFunctionOfTime holds a PiecewisePolynomial representing the angle through which the system has rotated, and the control error technically is the one for that internal PiecewisePolynomial.  However, QuaternionFunctionOfTime::update automatically does the correct update on its internal PiecewisePolynomial and recomputes the quaternion and its derivatives appropriately, in a way that is transparent to the user.)"

knelli2 commented:
@markscheel Added a `\note` to the docs of the Rotation control system detailing what you said and simplified the main description. Squashed directly.

Also rebased on develop.
